src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs

```diff
@@ -113,6 +114,22 @@ public AuthorizeFilter(string policy)
             }
 
             var effectivePolicy = Policy;
+                            
+            var mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
+            if (mvcOptions.CombineAuthorizeFilters)
+            {
+                // Combine all authorize filters into single effective policy that's only run on the first filter
+                if (context.Filters.First(f => f is AuthorizeFilter) != this)
+                {
+                    return;
+                }
+
+                var authorizeFilters = context.Filters.OfType<AuthorizeFilter>();
+                if (authorizeFilters.Count() > 1) {
+                    effectivePolicy = authorizeFilters.Select(s => s.Policy).Aggregate((p1, p2) => AuthorizationPolicy.Combine(p1, p2));
```

Comment:
```C#
AuthorizationPolicyBuilder builder;
for (var i = 0; i < context.Filters.Count; i++)
{
    if (ReferenceEquals(this, context.Filters[i]))
    {
        continue;
    }
    
    if (context.Filters[i] is AuthorizeFilter authorizeFilter)
    {
        builder = builder ?? new AuthorizationPolicyBuilder(effectivePolicy);
        builder.Combine(authorizeFilter.Policy);
    }
}

effectivePolicy = builder?.Build() ? effectivePolicy;
```

Reply:
That said, if this is a common-ish scenario, doing this in a filter created via ApplicationModel would avoid the allocations here entirely. Just something to consider.

Reply:
Is there any way we can cache this? I'm not stoked on the idea of all of this happening on each request

Reply:
We can cache if the policy provider is null just like the other policy code

Reply:
Related to this, I should add some custom policy provider tests to this to make sure this isn't horribly broken when using those

Reply:
Yeah, we should do per request work only if the user uses a custom IAuthorizationPolicyProvider